### DISCUSSION TITLE
Feature: Multi-cluster resources migration by karmada

### DISCUSSION THREAD

**Discussion by nivalume** [karmada-io/karmada#1097] [Ideas]
💡  Hi, guys, I want to discuss about the use of karmada to migrate resources, the general idea is as follows:

1. We have two clusters (such as Alibaba Cloud and Huawei Cloud) and want to migrate resources on Alibaba Cloud to Huawei Cloud cluster. We know that karmada can do multi-cluster resource distribution and scheduling.
2. Through karmada, we need to write a lot of propagation policies. But in our scenario, we need to migrate the resources of the entire cluster, so there is no need to write a propagation policy, only a very simple migration, such as `karmada migration foo-cluster` or write a migration `CRD`?

What really makes this migration difficult is:
1. Migration of stateful services, such as `Statefulset`
2. Hot/live migration, without interruption of service. (provided that the cluster to be migrated is unscheduler, by `karmada cordon foo-cluster`.
3. Global cluster migration, such as migrating from AWS to Huawei Cloud.

**Comment by RainbowMango** [comment]
I'm sure it's a reasonable feature. I suppose you can lead the effort :)

**Reply by nivalume** [reply]
Thanks, I am designing a proposal for this.

**Comment by jrkeen** [comment]
[velero](https://velero.io/) is a tool to back up and restore Kubernetes cluster resources and persistent volumes. Is it feasible for velero as a Karmada plugin?

**Reply by GitHubxsy** [reply]
Good idea. Deeper integration of Kamrada and velero

step1：Distribute velero-related crd through karmada for backup.
step2：Distribute velero-related crd through karmada for restore.
step3：...(I have no idea)..

**Comment by GitHubxsy** [comment]
This feature is so cool！！！

**Reply by nevermore-muyi** [reply]
This feature is so cool！！

**Comment by neujie** [comment]
So cool for it.
@jrkeen @algebra2k @RainbowMango @GitHubxsy @nevermore-muyi 
Maybe define a set of APIs in karmada is a good way.
I have an idea, and look forward to discussing with you.
![image](https://user-images.githubusercontent.com/8379762/147810924-108e0f65-f74f-4323-9df3-55a67fe614d6.png)

1. We define a set of CRs, such as Backup、Restore、Migration in Karmada;
2. Add a data controller in Karmada controller;
3. Create CRs, after successfully adding the K8S cluster;
4. Deploy data protection system, such as velero/portworx;
5. Deploy a CR controller(user define) to convert CRs to data protection system APIs.

We use velero as the default data protection system in karmada.

**Reply by nevermore-muyi** [reply]
Good idea. We can use karmada not only in kubernetes cluster management, but also in data management.

**Reply by GitHubxsy** [reply]
What are the capabilities of the `data controller` ?

**Reply by neujie** [reply]
1. Create CRs, after successfully adding the K8S cluster;
2. Sync operation status, such as Backup、Restore、Migration、Vault 、...  

As we discuss, its functions can continue to be enriched

**Reply by RainbowMango** [reply]
Impressive!!! @neujie can we have a talk at [the community meeting](https://docs.google.com/document/d/1y6YLVC-v7cmVAdbjedoyR5WL0-q45DBRXTvz5_I7bkA/edit#heading=h.vdip1san5drf)?

**Reply by nivalume** [reply]
Awesome!! your ideas are surprisingly consistent with ours. We have currently implemented `Backup`, `Restore` CR and related controllers on the internal production system.  We look forward to discussing with you and building it together.

**Reply by RainbowMango** [reply]
> We have currently implemented Backup, Restore CR and related controllers on the internal production system.

Glad to hear that!!! 

Happy New Year! By the way.
